### PR TITLE
[IR] Allow MDString in operand bundles

### PR DESCRIPTION
**Pull request by spavloff**
This change implements support of metadata strings in operand bundle values. It makes possible calls like:

    call void @some_func(i32 %x) [ "foo"(i32 42, metadata !"abc") ]

It requires some extension of the bitcode serialization. As SSA values and metadata are stored in different tables, there must be a way to distinguish them during deserialization. It is implemented by putting a special marker before the metadata index. The marker cannot be treated as a reference to any SSA value, so it unambiguously identifies metadata. It allows extending the bitcode serialization without breaking compatibility.

Metadata as operand bundle values are intended to be used in floating-point function calls. They would represent the same information as now is passed by the constrained intrinsic arguments.

### CODE REVIEW

File: llvm/lib/AsmParser/LLParser.cpp
Comment (arsenm) [inline llvm/lib/AsmParser/LLParser.cpp:6230]: Test the error case

File: llvm/lib/AsmParser/LLParser.cpp
Comment (arsenm) [inline llvm/lib/AsmParser/LLParser.cpp:3947]: Test error case

**Comment by nikic** [comment]
To check my understanding, the reason why we need the bitcode change is that we only encode the operand type if the value ID is larger than the instruction ID, which is never going to be the case for metadata, so we can't use the type to distinguish metadata? And the reason this is not a problem for normal call arguments is that it doesn't encode the types and instead takes them from the function type?

File: llvm/lib/AsmParser/LLParser.cpp
Comment (nikic) [inline llvm/lib/AsmParser/LLParser.cpp:6228]: Instead of adding this here, shouldn't we be calling parseMetadataAsValue() for metadata types in parseOptionalOperandBundles? That way we'd match how call arguments are parsed.

**Comment by spavloff** [comment]
> To check my understanding, the reason why we need the bitcode change is that we only encode the operand type if the value ID is larger than the instruction ID, which is never going to be the case for metadata, so we can't use the type to distinguish metadata?

Yes, the type index cannot be used for this purpose, because it is absent.

Also, there is no bitcode change here, everything that worked previously produces the same bit code. It is only the new feature (metadata in operande bundles) that require the new rule.

> And the reason this is not a problem for normal call arguments is that it doesn't encode the types and instead takes them from the function type?

Because previously deserializer always could expect either SSA value, or metadata, but not both.

File: llvm/lib/AsmParser/LLParser.cpp
Comment (spavloff) [reply llvm/lib/AsmParser/LLParser.cpp:6228]: This is determined by the logic of `parseOptionalOperandBundles`. It parses type then value:
https://github.com/llvm/llvm-project/blob/68c04b0ae62d8431d72d8b47fc13008002ee4387/llvm/lib/AsmParser/LLParser.cpp#L3205

Parsing value involves first parsing to `ValID`, then conversion to the value: https://github.com/llvm/llvm-project/blob/68c04b0ae62d8431d72d8b47fc13008002ee4387/llvm/lib/AsmParser/LLParser.cpp#L6274-L6279

Processing metadata values was implemented according to this logic.

Comment (spavloff) [reply llvm/lib/AsmParser/LLParser.cpp]: Changed `parseOptionalOperandBundles` accordingly.

**Review by nikic — APPROVED** [review]
LGTM, but please wait for a second approval.

File: llvm/test/Bitcode/operand-bundles.ll
Comment (arsenm) [inline llvm/test/Bitcode/operand-bundles.ll:59]: Should this also be tested in compatibility.ll

File: llvm/lib/AsmParser/LLParser.cpp
Comment (spavloff) [reply llvm/lib/AsmParser/LLParser.cpp]: This code is not present anymore.

File: llvm/lib/AsmParser/LLParser.cpp
Comment (spavloff) [reply llvm/lib/AsmParser/LLParser.cpp]: This code is not present anymore.

**Comment by spavloff** [comment]
Can someone else review theses changes? It needs two approvals to launch.

Thanks.

**Review by arsenm — APPROVED** [review]

File: llvm/lib/AsmParser/LLParser.cpp
Comment (paulwalker-arm) [inline llvm/lib/AsmParser/LLParser.cpp:3213]: This could be just:
```
else if (parseValue(Ty, Input, PFS))
  return true;
```

**Review by paulwalker-arm — APPROVED** [review]

**Comment by R-Goc** [comment]
This commit causes a warning during build time:
```
C:\Users\rysza\lib\llvm-project\llvm\lib\Bitcode\Reader\BitcodeReader.cpp(801,15): warning: comparison of integers of different signs: 'unsigned int' and 'llvm::bitc::MetadataOperandBundleValueMarker' [-Wsign-compare]
  801 |     if (ValID != bitc::OB_METADATA) {
      |         ~~~~~ ^  ~~~~~~~~~~~~~~~~~
```

**Comment by spavloff** [comment]
> This commit causes a warning during build time:
> 
> ```
> C:\Users\rysza\lib\llvm-project\llvm\lib\Bitcode\Reader\BitcodeReader.cpp(801,15): warning: comparison of integers of different signs: 'unsigned int' and 'llvm::bitc::MetadataOperandBundleValueMarker' [-Wsign-compare]
>   801 |     if (ValID != bitc::OB_METADATA) {
>       |         ~~~~~ ^  ~~~~~~~~~~~~~~~~~
> ```

@R-Goc Could you please check, if https://github.com/llvm/llvm-project/pull/113428 solves this problem?

**Comment by R-Goc** [comment]
Fixes it.

**Comment by spavloff** [comment]
Thanks!

**Pull request merged**
